# peterk87/nf-virontus: Usage

## Introduction

This pipeline performs read mapping and variant calling with [Minimap2] and [Medaka] with [Longshot] variant annotation. A consensus sequence is generated from major variants and variants that would not cause potential frameshift mutations using [Bcftools] with masking of low coverage depth regions with `N` characters.

Optionally, amplicon primers can be trimmed with [iVar] if a BED file of primer coordinates is supplied.

If read mapping against the SARS-CoV-2 reference genome Wuhan-Hu-1 ([MN908947.3](https://www.ncbi.nlm.nih.gov/nuccore/MN908947.3/)) is being performed, then [Pangolin] global SARS-CoV-2 lineage assignment will be performed.

## Running the pipeline

A typical command for running the pipeline with the reference genome SARS-CoV-2 MN908947.3 sequenced by the [ARTIC V3 protocol](https://github.com/artic-network/artic-ncov2019/tree/master/primer_schemes/nCoV-2019/V3) and samples and their reads specified in a sample sheet is as follows:

```bash
nextflow run peterk87/nf-virontus \
  -profile docker \
  --input my-sample-sheet.csv \
  --genome MN908947.3 \
  --artic_v3
```

This will launch the pipeline with the `docker` configuration profile. See below for more information about profiles.

Note that the pipeline will create the following files in your working directory:

```bash
work            # Directory containing the nextflow working files
results         # Finished results (configurable, see below)
.nextflow_log   # Log file from Nextflow
# Other nextflow hidden files, eg. history of pipeline runs and old logs.
```

## Help and usage info

The help and usage information for this pipeline can be displayed in your terminal with:

```bash
nextflow run peterk87/nf-virontus --help
```

Which should show a help and usage message like this:

```text
N E X T F L O W  ~  version 21.03.0-edge
Launching `main.nf` [grave_davinci] - revision: b1a24b7a67


------------------------------------------------------
  peterk87/nf-virontus v2.0.0
------------------------------------------------------

Typical pipeline command:

    nextflow run peterk87/nf-virontus --input samplesheet.csv --genome 'MN908947.3' --artic_v3 -profile docker

Input/output options
    --input                 [string]  Path to CSV, TSV, Excel XLSX or LibreOffice ODS file containing information about the samples you would like to analyse.
    --outdir                [string]  The output directory where the results will be saved.
    --save_cat_reads        [boolean] Option whether or not to save the concatenated reads FASTQ file for each sample to the output directory.

Reference genome options
    --genome                [string]  Name of viral reference genome. (e.g. 'MN908947.3')
    --fasta                 [string]  Path to FASTA genome file.
    --gff                   [string]  Full path to GFF annotation file.
    --primer_bed            [string]  A BED file containing primer coordinates. iVar is used to trim primer sequences after read alignment and before variant calling.
    --save_reference        [boolean] If generated by the pipeline save reference genome related files to the results folder.

SARS-CoV-2 Analysis Options
    --scov2                 [boolean] Convenience flag for setting `--genome 'MN908947.3'`, i.e. [SARS-CoV-2 Wuhan-Hu-1](https://www.ncbi.nlm.nih.gov/nuccore/MN908947.3/)
    --artic_v3              [boolean] Convenience flag for setting [ARTIC V3 primer BED file](https://github.com/artic-network/artic-ncov2019/blob/master/primer_schemes/nCoV-2019/V3/nCoV-2019.bed) for primer trimming.
    --freed                 [boolean] Convenience flag for setting 1200bp [Freed et al (2020)](https://academic.oup.com/biomethods/article/5/1/bpaa014/5873518) [primer BED file](https://zenodo.org/record/3897530) for primer trimming.

Variant calling options
    --medaka_model          [string]  [Medaka](https://github.com/nanoporetech/medaka) consensus model
    --medaka_chunk_len      [integer] [Medaka](https://github.com/nanoporetech/medaka) consensus chunk length of samples.
    --medaka_chunk_overlap  [integer] [Medaka](https://github.com/nanoporetech/medaka) consensus overlap of chunks
    --ivar_trim_noprimer    [boolean] This option unsets the '-e' parameter in 'ivar trim' to discard reads without primers. See [iVar docs for more info](https://andersen-lab.github.io/ivar/html/manualpage.html#autotoc_md16)
    --minor_allele_fraction [number]  Minor variant allele frequency/fraction.
    --major_allele_fraction [number]  Major variant allele frequency/fraction. Only major variant alleles are used for generating a consensus sequence.
    --low_coverage          [integer] Low coverage depth threshold. Consensus sequence positions with less than this coverage depth will be masked with `N`.

Phylogenetic Analysis Options
    --tree                  [boolean] Generate [IQ-TREE](http://www.iqtree.org/) phylogenetic tree from consensus sequences (and user-specified sequences).
    --iqtree_model          [string]  [IQ-TREE](http://www.iqtree.org/) substition model.
    --tree_extra_fasta      [string]  Optional sequences to include in phylogenetic tree construction along with consensus sequences generated from raw input sequence data.

Kraken2 analysis options
    --kraken2_db            [string]  Path to [Kraken2](https://ccb.jhu.edu/software/kraken2/) index/database.
    --skip_kraken2          [boolean] Skip [Kraken2](https://ccb.jhu.edu/software/kraken2/).

Slurm executor options
    --slurm_queue           [string]  Slurm queue or partition to run the pipeline on.
    --slurm_queue_size      [integer] Number of jobs Nextflow can queue simultaneously on the Slurm cluster.

Generic options
    --help                  [boolean] Display help text.
    --publish_dir_mode      [string]  Method used to save pipeline results to output directory.
    --monochrome_logs       [boolean] Do not use coloured log outputs.
    --multiqc_config        [string]  Custom config file to supply to MultiQC.
    --tracedir              [string]  Directory to keep pipeline Nextflow logs and reports.

Max job request options
    --max_cpus              [integer] Maximum number of CPUs that can be requested    for any single job.
    --max_memory            [string]  Maximum amount of memory that can be requested for any single job.
    --max_time              [string]  Maximum amount of time that can be requested for any single job.

------------------------------------------------------

If you use peterk87/nf-virontus for your analysis please cite:

* The pipeline
  TODO: add Zenodo DOI for pipeline

* The nf-core framework
  https://doi.org/10.1038/s41587-020-0439-x

* Software dependencies
  https://github.com/peterk87/nf-virontus/blob/master/CITATIONS.md

------------------------------------------------------
```

### Input/Output Options

Define where the pipeline should find input data and save output data.

#### `--input`

This workflow requires ONT Guppy basecalled (and demultiplexed) `.fastq` or `.fastq.gz` as input specified in a sample sheet table with 2 columns, 1 header row:

```text
sample,reads
SAMPLE_1,/path/to/barcoded-reads/barcode01
SAMPLE_2,/path/to/barcoded-reads/barcode02
SAMPLE_3,/path/to/concat-reads.fastq
SAMPLE_4,/path/to/concat-reads.fastq.gz
```

| Column   | Description  |
|----------|--------------|
| `sample` | Sample name. |
| `reads`  | Full path to directory containing FASTQ files (e.g. barcode demultiplexed reads generated by Guppy barcoder) or a FastQ file for all Nanopore reads belonging to a sample. FastQ files can be Gzip compressed with file extension `.fastq.gz` or `.fq.gz`. |

The reads path can be a directory containing barcode demultiplexed reads generated by [Guppy](https://nanoporetech.com/nanopore-sequencing-data-analysis). If multiple reads (`*.fastq` or `*.fastq.gz`) files are detected within a specified read directory, they are concatenated into a single Gzip compressed read file with the filename set as the user-specified that can be saved to the results output directory if `--save_cat_reads` is specified when running the workflow.

The format of the sample sheet file can be comma-separated values (CSV), tab-separated values (TSV), OpenDocument Spreadsheet Document (ODS) or Excel Spreadsheet (XLSX). For ODS and XLSX files, the first spreadsheet must contain the table starting at row 1 and column 1. Column headers are expected, but ignored.

> **NB:** It is **highly** recommended that the file paths to FASTQ files or directories containing FASTQ files be specified as absolute file paths (full file paths) rather than relative file paths. You can get the absolute file path for a file or directory using the [realpath](https://linux.die.net/man/3/realpath) Linux command-line utility (e.g. `realpath reads/*.fastq > fastq-filepaths.txt`).

#### `--outdir`

- Optional
- Type: string
- Default: `./results`

The output directory where the results will be saved.

#### `--save_cat_reads`

- Optional
- Type: boolean
- Default: `True`

Option whether or not to save the concatenated reads FASTQ file for each sample to the output directory.

### Reference genome options

Options for the reference genome indices used to align reads.

#### `--genome`

- Optional
- Type: string

Name of viral reference genome. (e.g. 'MN908947.3')

> **NOTE:** Only reference genomes specified in Nextflow configuration files are valid input.

#### `--fasta`

- Optional
- Type: string

Path to FASTA genome file.

> **NOTE:** If you have no genome reference available, the pipeline can build one using a FASTA file. This requires additional time and resources, so it's better to use a pre-build index if possible.

#### `--gff`

- Optional
- Type: string

Full path to GFF annotation file.

#### `--primer_bed`

- Optional
- Type: string

A BED file containing primer coordinates. iVar is used to trim primer sequences after read alignment and before variant calling.

> **NOTE:** iVar uses the primer positions relative to the viral genome supplied in this file to soft clip primer sequences from a coordinate sorted BAM file. The file must be in BED format as highlighted below:

```text
MN908947.3 30 54 nCoV-2019_1_LEFT 60 -
MN908947.3 385 410 nCoV-2019_1_RIGHT 60 +
MN908947.3 320 342 nCoV-2019_2_LEFT 60 -
MN908947.3 704 726 nCoV-2019_2_RIGHT 60 +
```

#### `--save_reference`

- Optional
- Type: boolean

If generated by the pipeline save reference genome related files to the results folder.

### SARS-CoV-2 Analysis Options

SARS-CoV-2 specific analysis options and convenience flags.

#### `--scov2`

- Optional
- Type: boolean

Convenience flag for setting `--genome 'MN908947.3'`, i.e. [SARS-CoV-2 Wuhan-Hu-1](https://www.ncbi.nlm.nih.gov/nuccore/MN908947.3/)

#### `--artic_v3`

- Optional
- Type: boolean

Convenience flag for setting [ARTIC V3 primer BED file](https://github.com/artic-network/artic-ncov2019/blob/master/primer_schemes/nCoV-2019/V3/nCoV-2019.bed) for primer trimming.

#### `--freed`

- Optional
- Type: boolean

Convenience flag for setting 1200bp [Freed et al (2020)](https://academic.oup.com/biomethods/article/5/1/bpaa014/5873518) [primer BED file](https://zenodo.org/record/3897530) for primer trimming.

### Variant calling options

Various options for the variant calling branch of the pipeline.

#### `--medaka_model`

- Optional
- Type: string
- Default: `r941_min_high_g360`

[Medaka](https://github.com/nanoporetech/medaka) consensus model

#### `--medaka_chunk_len`

- Optional
- Type: integer
- Default: `800`

[Medaka](https://github.com/nanoporetech/medaka) consensus chunk length of samples.

#### `--medaka_chunk_overlap`

- Optional
- Type: integer
- Default: `400`

[Medaka](https://github.com/nanoporetech/medaka) consensus overlap of chunks

#### `--ivar_trim_noprimer`

- Optional
- Type: boolean

This option unsets the '-e' parameter in 'ivar trim' to discard reads without primers. See [iVar docs for more info](https://andersen-lab.github.io/ivar/html/manualpage.html#autotoc_md16)

#### `--minor_allele_fraction`

- Optional
- Type: number
  - Maximum: 1.0
- Default: `0.25`

Minor variant allele frequency/fraction.

#### `--major_allele_fraction`

- Optional
- Type: number
  - Maximum: 1.0
- Default: `0.75`

Major variant allele frequency/fraction. Only major variant alleles are used for generating a consensus sequence.

#### `--low_coverage`

- Optional
- Type: integer
- Default: `10`

Low coverage depth threshold. Consensus sequence positions with less than this coverage depth will be masked with `N`.

### Phylogenetic Analysis Options

Options for performing phylogenetic analysis with [MAFFT](https://mafft.cbrc.jp/alignment/software/) multiple sequence alignment (MSA) of consensus sequences (and optional user-specified sequences) and [IQ-TREE](http://www.iqtree.org/) maximum-likelihood phylogenetic tree inference. Tree visualization will be output in MultiQC report.

#### `--tree`

- Optional
- Type: boolean

Generate [IQ-TREE](http://www.iqtree.org/) phylogenetic tree from consensus sequences (and user-specified sequences).

#### `--iqtree_model`

- Optional
- Type: string
- Default: `GTR`

[IQ-TREE](http://www.iqtree.org/) substition model.

#### `--tree_extra_fasta`

- Optional
- Type: string

Optional sequences to include in phylogenetic tree construction along with consensus sequences generated from raw input sequence data.

### Kraken2 analysis options

Options for [Kraken2](https://ccb.jhu.edu/software/kraken2/) taxonomic classification and host subtraction analysis.

#### `--kraken2_db`

- Optional
- Type: string

Path to [Kraken2](https://ccb.jhu.edu/software/kraken2/) index/database.

#### `--skip_kraken2`

- Optional
- Type: boolean

Skip [Kraken2](https://ccb.jhu.edu/software/kraken2/).

#### `--subtract_host`

- Optional
- Type: boolean

Subtract host (and any other classified) reads prior to analysis.

### Slurm executor options

Options for executing this pipeline on a Slurm scheduler for an HPC cluster. The Slurm queue (or partition) must be specified in order to run the pipeline on Slurm cluster.

#### `--slurm_queue`

- Optional
- Type: string

Slurm queue or partition to run the pipeline on.

> **NOTE:** This will be specific to the cluster you're using! Please check with the available queues with `$ sinfo` or check with your cluster admins to see which queue to use.

#### `--slurm_queue_size`

- Optional
- Type: integer
  - Minimum: 1
- Default: `100`

Number of jobs Nextflow can queue simultaneously on the Slurm cluster.

> **NOTE:** Please respect others using the cluster. The Slurm cluster is a likely a public resource so please specify a reasonable value.

### Generic options

Less common options for the pipeline, typically set in a config file.

#### `--help`

- Optional
- Type: boolean

Display help text.

#### `--publish_dir_mode`

- Optional
- Type: string
- Default: `copy`

Method used to save pipeline results to output directory.

> **NOTE:** The Nextflow `publishDir` option specifies which intermediate files should be saved to the output directory. This option tells the pipeline what method should be used to move these files. See [Nextflow docs](https://www.nextflow.io/docs/latest/process.html#publishdir) for details.

#### `--monochrome_logs`

- Optional
- Type: boolean

Do not use coloured log outputs.

> **NOTE:** Set to disable colourful command line output and live life in monochrome.

#### `--multiqc_config`

- Optional
- Type: string

Custom config file to supply to MultiQC.

#### `--tracedir`

- Optional
- Type: string
- Default: `${params.outdir}/pipeline_info`

Directory to keep pipeline Nextflow logs and reports.

#### `--name`

- Optional
- Type: string

Custom name for pipeline execution. If not specified, then a random name will be generated by Nextflow.

#### `--container`

- Optional
- Type: string
- Default: `peterk87/nf-virontus:2.0.0`

Container to use for execution of pipeline.

> **NOTE:** This should not be changed unless you know what you're doing.

### Max job request options

Set the top limit for requested resources for any single job.

#### `--max_cpus`

- Optional
- Type: integer
- Default: `16`

Maximum number of CPUs that can be requested    for any single job.

> **NOTE:** Use to set an upper-limit for the CPU requirement for each process. Should be an integer e.g. `--max_cpus 1`

#### `--max_memory`

- Optional
- Type: string
- Default: `128.GB`

Maximum amount of memory that can be requested for any single job.

> **NOTE:** Use to set an upper-limit for the memory requirement for each process. Should be a string in the format integer-unit e.g. `--max_memory '8.GB'`

#### `--max_time`

- Optional
- Type: string
- Default: `240.h`

Maximum amount of time that can be requested for any single job.

> **NOTE:** Use to set an upper-limit for the time requirement for each process. Should be a string in the format integer-unit e.g. `--max_time '2.h'`

### Updating the pipeline

When you run the above command, Nextflow automatically pulls the pipeline code from GitHub and stores it as a cached version. When running the pipeline after this, it will always use the cached version if available - even if the pipeline has been updated since. To make sure that you're running the latest version of the pipeline, make sure that you regularly update the cached version of the pipeline:

```bash
nextflow pull peterk87/nf-virontus
```

### Reproducibility

It's a good idea to specify a pipeline version when running the pipeline on your data. This ensures that a specific version of the pipeline code and software are used when you run your pipeline. If you keep using the same tag, you'll be running the same version of the pipeline, even if there have been changes to the code since.

First, go to the [peterk87/nf-virontus releases page](https://github.com/peterk87/nf-virontus/releases) and find the latest version number - numeric only (eg. `2.0.0`). Then specify this when running the pipeline with `-r` (one hyphen) - eg. `-r 2.0.0`.

This version number will be logged in reports when you run the pipeline, so that you'll know what you used when you look back in the future.

## Core Nextflow arguments

> **NB:** These options are part of Nextflow and use a _single_ hyphen (pipeline parameters use a double-hyphen).

### `-profile`

Use this parameter to choose a configuration profile. Profiles can give configuration presets for different compute environments.

Several generic profiles are bundled with the pipeline which instruct the pipeline to use software packaged using different methods (Docker, Singularity, Conda) - see below.

> We highly recommend the use of Docker or Singularity containers for full pipeline reproducibility, however when this is not possible, Conda is also supported. When performing SARS-CoV-2 analysis with this workflow, the use of Docker or Singularity containers is required to run the latest version of [Pangolin].

<!-- TODO: dynamically load configs for this pipeline? 
The pipeline also dynamically loads configurations from [https://github.com/nf-core/configs](https://github.com/nf-core/configs) when it runs, making multiple config profiles for various institutional clusters available at run time. For more information and to see if your system is available in these configs please see the [nf-core/configs documentation](https://github.com/nf-core/configs#documentation). -->

Note that multiple profiles can be loaded, for example: `-profile test,docker` - the order of arguments is important!
They are loaded in sequence, so later profiles can overwrite earlier profiles.

If `-profile` is not specified, the pipeline will run locally and expect all software to be installed and available on the `PATH`. This is _not_ recommended.

- `docker`
  - A generic configuration profile to be used with [Docker](https://docker.com/)
  - Pulls software from Docker Hub: [`peterk87/nf-virontus`](https://hub.docker.com/r/peterk87/nf-virontus/)
- `singularity`
  - A generic configuration profile to be used with [Singularity](https://sylabs.io/docs/)
  - Pulls software from Docker Hub: [`peterk87/nf-virontus`](https://hub.docker.com/r/peterk87/nf-virontus/)
- `conda`
  - Please only use Conda as a last resort i.e. when it's not possible to run the pipeline with Docker or Singularity.
  - A generic configuration profile to be used with [Conda](https://conda.io/docs/)
  - Pulls most software from [Bioconda](https://bioconda.github.io/)
- `test`
  - A profile with a complete configuration for automated testing
  - Includes links to test data so needs no other parameters

### `-resume`

Specify this when restarting a pipeline. Nextflow will used cached results from any pipeline steps where the inputs are the same, continuing from where it got to previously.

You can also supply a run name to resume a specific run: `-resume [run-name]`. Use the `nextflow log` command to show previous run names.

### `-c`

Specify the path to a specific config file (this is a core Nextflow command). See the [nf-core website documentation](https://nf-co.re/usage/configuration) for more information.

#### Custom resource requests

Each step in the pipeline has a default set of requirements for number of CPUs, memory and time. For most of the steps in the pipeline, if the job exits with an error code of `143` (exceeded requested resources) it will automatically resubmit with higher requests (2 x original, then 3 x original). If it still fails after three times then the pipeline is stopped.

Whilst these default requirements will hopefully work for most people with most data, you may find that you want to customise the compute resources that the pipeline requests. You can do this by creating a custom config file. For example, to give the workflow process `MEDAKA_LONGSHOT` 16GB of memory and 16 CPUs, you could use the following config:

```nextflow
process {
  withName: MEDAKA_LONGSHOT {
    cpus = 16
    memory = 16.GB
  }
}
```

See the main [Nextflow documentation](https://www.nextflow.io/docs/latest/config.html) for more information.

<!-- TODO: uncomment when pipeline added to nf-core

If you are likely to be running `nf-core` pipelines regularly it may be a good idea to request that your custom config file is uploaded to the `nf-core/configs` git repository. Before you do this please can you test that the config file works with your pipeline of choice using the `-c` parameter (see definition below). You can then create a pull request to the `nf-core/configs` repository with the addition of your config file, associated documentation file (see examples in [`nf-core/configs/docs`](https://github.com/nf-core/configs/tree/master/docs)), and amending [`nfcore_custom.config`](https://github.com/nf-core/configs/blob/master/nfcore_custom.config) to include your custom profile.

If you have any questions or issues please send us a message on [Slack](https://nf-co.re/join/slack) on the [`#configs` channel](https://nfcore.slack.com/channels/configs).
-->

### Running in the background

Nextflow handles job submissions and supervises the running jobs. The Nextflow process must run until the pipeline is finished.

The Nextflow `-bg` flag launches Nextflow in the background, detached from your terminal so that the workflow does not stop if you log out of your session. The logs are saved to a file.

Alternatively, you can use `screen` / `tmux` or similar tool to create a detached session which you can log back into at a later time.
Some HPC setups also allow you to run nextflow within a cluster job submitted your job scheduler (from where it submits more jobs).

#### Nextflow memory requirements

In some cases, the Nextflow Java virtual machines can start to request a large amount of memory.
We recommend adding the following line to your environment to limit this (typically in `~/.bashrc` or `~./bash_profile`):

```bash
NXF_OPTS='-Xms1g -Xmx4g'
```

[Bcftools]: https://samtools.github.io/bcftools/bcftools.html
[Centrifuge]: https://ccb.jhu.edu/software/centrifuge/manual.shtml
[Conda]: https://conda.io/
[Docker]: https://www.docker.com/
[IQ-TREE]: http://www.iqtree.org/
[iVar]: https://github.com/andersen-lab/ivar
[Kraken2]: https://ccb.jhu.edu/software/kraken2/
[Longshot]: https://www.nature.com/articles/s41467-019-12493-y
[MAFFT]: https://mafft.cbrc.jp/alignment/software/
[Matplotlib]: https://matplotlib.org/
[Medaka]: https://github.com/nanoporetech/medaka
[Minimap2]: https://github.com/lh3/minimap2
[Mosdepth]: https://github.com/brentp/mosdepth
[MultiQC]: http://multiqc.info
[Nextflow]: https://www.nextflow.io
[Pangolin]: https://github.com/cov-lineages/pangolin/
[pigz]: https://www.zlib.net/pigz/
[Samtools]: https://www.htslib.org/
[seaborn]: https://seaborn.pydata.org/
[Singularity]: https://sylabs.io/guides/3.5/user-guide/
[SnpEff]: https://pcingola.github.io/SnpEff/
[SnpSift]: https://pcingola.github.io/SnpEff/ss_introduction/
[vcf_consensus_builder]: https://github.com/peterk87/vcf_consensus_builder
